### PR TITLE
New script to help with automating major steps of evaluating a PR

### PR DESCRIPTION
**Pull request by jwrober**
Not related to an existing issue.

### CODE REVIEW

File: scripts/eval-pr [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  echo "-- Getting a .diff of PR $pull_request."
+  echo ""
+  wget https://github.com/longturn/freeciv21/pull/$pull_request.diff -O pr$pull_request.diff
+
+  echo ""
+  echo "-- Applying the .diff via patch."
+  echo ""
+  patch -p1 < pr$pull_request.diff
+  rm pr$pull_request.diff
+  rm -Rf build
```
Comment (lmoureaux) [inline scripts/eval-pr:68]: Noo you nuked my build directory!

**Review by lmoureaux — COMMENTED** [review]
Very useful idea. For some reason the script failed to create the branch in my case, and thus cleanup also failed. The `-a` option is strange, it rebuilds everything from scratch even if `-s` was done before.

I think I'm going to use this from an entirely separate checkout of the source code, so I can run a test build with `nice ./eval-pr` and it won't interfere with development. Would be useful to have a `-j` option to limit the number of compiler processes and thus memory usage. Configuring the build folder with `ccache` would potentially speed up the build immensely.

**Comment by pranavtifr** [comment]
I think the build needs to happen in a different directory to not interfere with the current workflow of the dev. Install and package is maybe not needed just to evaluate a PR. 

Maybe do builds in folder like build_$pull_request so that people can checkout and evaluate multiple PRs at once.

File: scripts/eval-pr [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    echo "Error: Attempting cleanup on incorrect pull request number."
+    echo "The current eval is on branch: $branch"
+    echo ""
+    exit 255
+  fi
+
+  echo "-- Stashing Changes and checking out master branch."
+  echo ""
+  git stash
+  git checkout master
```
[the remark this answers is not in the document]
Comment (jwrober) [reply scripts/eval-pr:142]: Idea(s) on how to improve this or maybe best to throw and error and force the human to correct?

File: scripts/eval-pr [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+  base_setup
+
+  echo ""
+  echo "-- Building."
+  echo ""
+  cmake -S . -B build -G Ninja -DCMAKE_INSTALL_PREFIX=$PWD/build/install -DCMAKE_BUILD_TYPE=Debug
+  cmake --build build
+  cmake --build build --target install
+  cmake --build build --target package
```
[the remark this answers is not in the document]
Comment (jwrober) [reply scripts/eval-pr:85]: Good catch. I can do an OS check first. I still this it a good idea to run the package test however.

File: scripts/eval-pr [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  echo "-- Starting an Eval Process for PR $pull_request."
+  echo ""
+
+  base_setup
+
+  echo ""
+  echo "-- Building."
+  echo ""
+  cmake -S . -B build -G Ninja -DCMAKE_INSTALL_PREFIX=$PWD/build/install -DCMAKE_BUILD_TYPE=Debug
+  cmake --build build
```
[the remark this answers is not in the document]
Comment (jwrober) [reply scripts/eval-pr:83]: I'm going to use @pranavtifr idea and do a special purpose build directory that won't overwrite anything in use already

File: scripts/eval-pr [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  echo ""
+  echo "-- Starting an Eval Process for PR $pull_request."
+  echo ""
+
+  base_setup
+
+  echo ""
+  echo "-- Building."
+  echo ""
+  cmake -S . -B build -G Ninja -DCMAKE_INSTALL_PREFIX=$PWD/build/install -DCMAKE_BUILD_TYPE=Debug
```
[the remark this answers is not in the document]
Comment (jwrober) [reply scripts/eval-pr:82]: See note above.

File: scripts/eval-pr (hunk 1 above)
Comment (jwrober) [reply scripts/eval-pr:68]: Yea I like @pranavtifr idea to do a special purpose build directory to leave the standard one alone.

**Comment by jwrober** [comment]
If ok, I suggest a squash and merge.

**Comment by pranavtifr** [comment]
One last suggestion so that we dont have to come back to this later. 

Would it possible to put the whole thing into a loop so that, multiple PRs are checked and built ? I would find it convenient where I can just ask the script to build multiple PRs and come have a lot at it later when the build process is done.

**Comment by lmoureaux** [comment]
> Would it possible to put the whole thing into a loop so that, multiple PRs are checked and built ? I would find it convenient where I can just ask the script to build multiple PRs and come have a lot at it later when the build process is done.

This would be rather easy, also solving the CodeFactor issue saying that the options loop gets executed only once.

**Review by lmoureaux — APPROVED** [review]
No objection to merging

**Pull request merged**